### PR TITLE
drivers: adc: STM32: Support Oversampling on all H7 devices

### LINKED ISSUE #38606: drivers: adc: stm32h7: Oversampling only supported for ADC3
**Describe the bug**
This [PR](https://github.com/zephyrproject-rtos/zephyr/pull/36714/) provided the ability to do ADC oversampling, which then also added support for the H7 in this [PR](https://github.com/zephyrproject-rtos/zephyr/pull/37392). This will however not set the ratio correctly.

According to the [Zephyr ADC docs](https://docs.zephyrproject.org/latest/reference/peripherals/adc.html#c.adc_sequence.oversampling) the value `sequence->oversampling` should be `2^N`. This means if you require a ratio of 256 you would define `sequence->oversampling=8`. But using ADC1 or ADC2 with the current code if you set `sequence->oversampling=8` it will result in a ratio of 7, and not 256. It requires the ratio in a different format than for ADC3 for which it seems this driver was written.

**For ADC1 and ADC2:**
> OSVR[9:0]: Oversampling ratio
This bitfield is set and cleared by software to define the oversampling ratio.
0: 1x (no oversampling)
1: 2x
2: 3x
...
1023: 1024x

**For ADC3:** 
> OVSR[2:0]: Oversampling ratio
This bitfield is set and cleared by software to define the oversampling ratio.
000: 2x
001: 4x
010: 8x
011: 16x
100: 32x
101: 64x
110: 128x
111: 256x

If using ADC1 or ADC2 I would recommend simply stating:
```
uint16_t oversampling_ratio = (1<<sequence->oversampling) - 1;
...
LL_ADC_ConfigOverSamplingRatioShift(adc, oversampling_ratio, ...);
```

### PR DESCRIPTION
**Pull request by heinwessels**
Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/38606 and another bug mentioned in the comments.

Signed-off-by: Hein Wessels <hein.wessels@nobleo.nl>

---

Edit: All H7 devices has ratio set 0->1023 for ADC1 and ADC2, not only when `ADC_VER_V5_V90`.

### CODE REVIEW

**Review by JvanDooren — APPROVED** [review]

**Comment by heinwessels** [comment]
Now handles a [bug](https://github.com/STMicroelectronics/STM32CubeH7/issues/177) in the HAL Libraries correctly by manually setting modifying registers.

File: drivers/adc/adc_stm32.c [hunk 1]
```diff
@@ -422,51 +422,46 @@ static int start_read(const struct device *dev,
...
+};
+
+	uint32_t oversampling_ratio = sequence->oversampling;
+#if defined(CONFIG_SOC_SERIES_STM32H7X) && defined(ADC_VER_V5_V90)
+	if (adc != ADC3) {
+		oversampling_ratio = 1<<sequence->oversampling;
+	}
+#endif
+	if (sequence->oversampling == 0) {
 		LL_ADC_SetOverSamplingScope(adc, LL_ADC_OVS_DISABLE);
```
Comment (JvanDooren) [inline drivers/adc/adc_stm32.c:446]: So this works for ADC3 also? At least the bits 10 and 9 are the same in cfg2 register for adc1/2/3, so probably yes.

File: drivers/adc/adc_stm32.c [hunk 2]
```diff
@@ -422,51 +422,46 @@ static int start_read(const struct device *dev,
...
-						    LL_ADC_OVS_SHIFT_RIGHT_3);
-		break;
-	case 4:
-		LL_ADC_SetOverSamplingScope(adc, LL_ADC_OVS_GRP_REGULAR_CONTINUED);
-		LL_ADC_ConfigOverSamplingRatioShift(adc, LL_ADC_OVS_RATIO_16,
-						    LL_ADC_OVS_SHIFT_RIGHT_4);
-		break;
-	case 5:
+	} else if (sequence->oversampling <= 8) {
 		LL_ADC_SetOverSamplingScope(adc, LL_ADC_OVS_GRP_REGULAR_CONTINUED);
```
Comment (JvanDooren) [inline drivers/adc/adc_stm32.c:448]: So this works for ADC3 also? At least the bits 10 and 9 are the same in cfg2 register for adc1/2/3, so probably yes.

**Review by JvanDooren — CHANGES_REQUESTED** [review]

File: drivers/adc/adc_stm32.c (hunk 2 above)
Comment (heinwessels) [reply drivers/adc/adc_stm32.c:448]: Yes, the settings are in the same place in `CFG2`.

File: drivers/adc/adc_stm32.c (hunk 1 above)
Comment (heinwessels) [reply drivers/adc/adc_stm32.c:446]: Yes, the settings are in the same place in `CFG2`.

**Review by JvanDooren — APPROVED** [review]

**Review by royvandam — APPROVED** [review]

**Pull request merged**